src/resources/s3.js

```diff
@@ -542,6 +566,9 @@ class S3Bucket { // eslint-disable-line padded-blocks
 			} else {
 				await this._deleteBucketEncryption(bucket.metadata.name);
 			}
+			if (versioningConfiguration) {
+				await this._putVersioningConfiguration(bucket.metadata.name, versioningConfiguration);
+			}
```

Comment:
Our goal with the aws-resource-service is to have the bucket resource describe the state of the bucket, so whatever we do needs to ensure that that goal stays intact.

Reading the comment and the docs indeed means we cannot go back to 'unversioned', but we could go to 'suspended'. I think my ideal situation would the following:
1. If the bucket is never configured for versioning, then we do not automatically configure it ('unversioned' stays 'unversioned')
2. If the bucket is 'unversioned', and now someone sets the configuration to either "Enabled" or "Suspended", then that's just fine, and the bucket is now 'versioned'.
3. If the bucket is 'versioned', then you cannot remove the version configuration from it, i.e. if there was a version configuration, then it not being there in an update is an error.
4. If the bucket is 'versioned' you can update it between 'Enabled'/'Suspended' as much as you want, it will stay 'versioned'

Code-wise this means in the update case we need to query the version status when there is no version configuration, and specifically check whether the bucket is 'unversioned' (no configuration is just fine) or 'versioned' (no configuration is an error).

Does that make sense?

Reply:
Looking at https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getBucketVersioning-property and the REST variant: I suspect AWS in JS will return an empty data object for 'unversioned' buckets.

Reply:
> If the bucket is 'versioned', then you cannot remove the version configuration from it, i.e. if there was a version configuration, then it not being there in an update is an error.

This will lead to a breaking change.
Scenario:
1. User manually activate/deactivated versioning -> in the AWS console this looks the same as versioning having never been configured
2. New version of `kubernetes-aws-resource-service` separates between "never configured" vs "suspended"
3. Old `kubernetes-aws-resource-service` config (without versioning config) will now be rejected by by new `kubernetes-aws-resource-service`

Reply:
Thinking further, it seems to be boil down what "not configured" means. Kind of `undefined` vs `null` :)
For example, if we remove the definition for a bucket from the config -> would `kubernetes-aws-resource-service` delete the bucket (`null`) or consider to be no longer in charge of it (`undefined`)?

Reply:
> This will lead to a breaking change. [...]

Right. Now one could raise how relevant that is, after all the user has decided to poke things through external means. Generally we handle that just fine, but will attempt to restore the resource to our own liking when we run the next update/sync. That will now fail as you describe. I'm a bit on the fence here, I can perfectly see that being ok, and the user now needing to provide an explicit versioning configuration here, paying for their fault to follow the procedures.

I can also see how this might be annoying, especially since the distinction is somewhat academical, and has no practical meaning[*]. Assuming one goes with that, then I suspect we would have to always configure the versioning to either 'Enabled' (if the user said so) or 'Suspended' (if they didn't). We could potentially avoid configuring versioning in the case where the user didn't specify anything, *and* the current versioning is "unconfigured", to at least leave things as they are. If the versioning state is anything but unconfigured we would in the update case always set the value.

> For example, if we remove the definition for a bucket from the config -> would kubernetes-aws-resource-service delete the bucket (null) or consider to be no longer in charge of it (undefined)?

If a resource is removed from Kubernetes, then the resource service "generally" will delete the resource also in AWS. "Generally" though, because for S3 buckets we chicken out: https://github.com/Collaborne/kubernetes-aws-resource-service/blob/9c6fc181ca46281d0377b83609209313e44f4059/src/resources/s3.js#L527-L535 is called, but this will fail when the bucket isn't empty. And we do not try to empty it first :)

---
[*]: Except: A bucket with versioning in suspended state might have a lot of versions stored in it, and you need to pay for those bytes. A unattent operator investigating storage costs might only look at the console, and then fail to recognize that the bucket actually has versions.

Reply:
This has been implemented by https://github.com/Collaborne/kubernetes-aws-resource-service/pull/59/commits/38840b655a88035c821c399dc937bc4b5d72c410